lightning/src/ln/channel.rs

```diff
@@ -3134,12 +3134,24 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 	}
 
 	/// Called by channelmanager based on chain blocks being connected.
-	/// Note that we only need to use this to detect funding_signed, anything else is handled by
-	/// the channel_monitor.
+	/// We need to use this to detect funding_signed and outgoing HTLC timed out before we were able
+	/// to commit them on remote commitment tx, anything else is handled by the channel_monitor.
 	/// In case of Err, the channel may have been closed, at which point the standard requirements
 	/// apply - no calls may be made except those explicitly stated to be allowed post-shutdown.
 	/// Only returns an ErrorAction of DisconnectPeer, if Err.
-	pub fn block_connected(&mut self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[u32]) -> Result<Option<msgs::FundingLocked>, msgs::ErrorMessage> {
+	pub fn block_connected(&mut self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[u32]) -> Result<(Option<msgs::FundingLocked>, Vec<(HTLCSource, PaymentHash, u64)>), msgs::ErrorMessage> {
+		let mut timed_out_htlcs = Vec::new();
+		self.holding_cell_htlc_updates.retain(|htlc_update| {
+			match htlc_update {
+				&HTLCUpdateAwaitingACK::AddHTLC { ref payment_hash, ref source, ref cltv_expiry, ref amount_msat, .. } => {
+					if cltv_expiry <= &height {
+						timed_out_htlcs.push((source.clone(), payment_hash.clone(), *amount_msat));
+						false
```

Comment:
Might be good to add a test for this case, to ensure holding cell HTLCs are timed out properly?

Reply:
Done! Caught a few bugs along the way, of course :/.